File: PreorderTraversal.java
package com.wrh.leetcode;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class PreorderTraversal {
	List<Integer> res=new ArrayList<Integer>();
	public List<Integer> preorderTraversal(TreeNode root) {
        if(root==null){
        	return res;
        }
        Stack<TreeNode> stack=new Stack<TreeNode>();
        stack.push(root);
        while(!stack.isEmpty()){
        	TreeNode node=stack.pop();
        	res.add(node.val);
        	if(node.right!=null){
        		stack.push(node.right);
        	}
        	if(node.left!=null){
        		stack.push(node.left);
        	}
        }
		return res;
    }
}

File: preorderTraversal.cpp


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
/**
 * Return an array of size *returnSize.
 * Note: The returned array must be malloced, assume caller calls free().
 */
void preorderTraversalHelper(struct TreeNode* root,int *arr,int* returnSize){
	if(root==NULL){
		return ;
	}
	arr[*returnSize]=root->val;
	(*returnSize)++;
	preorderTraversalHelper(root->left,arr,returnSize);
	preorderTraversalHelper(root->right,arr,returnSize);	
}
int* preorderTraversal(struct TreeNode* root, int* returnSize) {
	int *res=(int *)malloc(N*sizeof(int));
	if(res==NULL){
		exit(EXIT_FAILURE);
	}
	*returnSize=0;
    preorderTraversalHelper(root,res,returnSize);
    return res;
}
